### ISSUE TITLE
Be more conservative about migration?

### ISSUE THREAD

**Issue by ekr** [quicwg/base-drafts#2143]
I've been reading through the migration mechanism and it seems to me
that assuming that a packet from a given IP address is evidence of
migration creates a pretty easy DoS attack from an on-path attacker.
Consider the case where we have a quiescent connection between A and B
and an on-path attacker and then B sends a packet to A.

In the diagram below, the notation [A:B]: Z -> means a packet with a
source addr of A, a dest addr of B, and a payload of Z

```
A                       Attacker (Z)                  B

                                         <- [B:A]: Data
                     <- [B:A]: Data
[A:B]: ACK ->
                        [Z:B]: ACK ->

                                       X <- [B:Z]: Data
                                       X <- [B:Z]: Probe
```
                                       
As long as A doesn't try to send a data packet, B will continue to
think that A is at Z for the duration of the validation timer
(currently 3 RTO?) [0]. That's not the worst thing in the world, but
it's a pretty good cost/benefit for an on-path attacker who only has
to forge one packet; I'm not sure that that's a great design tradeoff
for a protocol where we're otherwise trying to stop on-path attacker
from interfering with the connection after the handshake completes.

The reason this works is the receipt of a packet from addr X is a
trigger for abandoning the previous path and trying migration. Another
time this can happen is at the tail of a client transmission,
though that will presumably recover faster.

At a high level, it seems like there are two major types of
migration:

1. Unknown migration (e.g., NAT re-binding)
2. Known migration (e.g., transferring from WiFi to 3G)

In the latter case, we can have the client give notice that it's
migrating, even if it doesn't know it's visible IP address, and have
the server refuse to change peer addresses unless it sees that
indication, so it's just the former case we have to deal with, and as
above. The present design does that but at the cost of creating the
situation above.

I wonder if we should be more conservative about initiating migration.
I don't have a fully worked out proposal, but here are some ideas:

- Send packets to both the original and new addresses during the
  validation period.
  
- If you receive an apparent migration that looks like a rebinding
  (e.g., same address, different port) within X seconds of a previous
  packet, then probe first (or in parallel as above).

- Have the client send some packet that indicates it thinks it's
  migrating (this seems tricky to get right)

[0] Note that if we were to allow migration on the client and the
server, even worse attacks are possible.

**Comment by mikkelfj** [comment]
When you say on-path, do you mean observer? I.e. with the ability to race a packet, or to drop or replace a packet.

I suppose a race will not be successful in this attack.

So the effectiveness on the attack builds on it being cheaper to confuse the connection rather than to actively drop packets. I'm not sure that is a strong case for additional complexity. However, a related attack might be used for amplification attacks by changing the IP to that of the indented victim.

Attacks that can be instigated by observers are much more severe because they are far easier to achieve, especially undetected.

**Comment by ekr** [comment]
For the purposes of this discussion, I'm talking about an attacker who can drop-and-replace, but I'm not sure why you think a race won't work, as it depends quite a bit on the network topoligy

**Comment by mikkelfj** [comment]
To prevent a race from being successful the probe needs to checksum the visible address and have that checksum be reflected back by the client rather than just ACK'ing the probe. Once the reflected checksum is seen the path must be valid.

For true on path, the attacker can drop packets from client to server to maintain the illusion of a different client address, but then the probe will never be reflected correctly. It will eventually time out.

BTW: the description on notation is confusing, mixing [A:B] with [X:Y] and sometimes having Z as payload and sometimes as an address.

**Comment by mikkelfj** [comment]
No that is not sufficient. The attacker can just race the reflected probe checksum.

**Comment by martinthomson** [comment]
What we have in practice is this:
```
A                       Attacker (Z)                  B

                                         <- [B:A]: Data
                     <- [B:A]: Data
[A:B]: ACK ->
                        [Z:B]: ACK ->

                                       X <- [B:Z]: Data
                                         <- [B:A]: Probe
                                       X <- [B:Z]: Probe
```

Provided the attacker isn't on path (where they could deny service anyway), the result is that the migration will fail as soon as the probe response arrives.

**Comment by ekr** [comment]
Why is it probing B->A?

**Comment by martinthomson** [comment]
Because we [told it to](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#off-path-forward): 

> In response to an apparent migration, endpoints MUST validate the previously active path using a PATH_CHALLENGE frame.

**Comment by mikkelfj** [comment]
> That is, I think that we have all of what you suggest already, with the exception of the message saying "I think that I'm migrating". I don't know how to do that. We don't really create any signals that A might use to learn that it is migrating.

If you see a new CID on a packet that is not Initial, you are probably migrating. You can win races by copying and spoofing source address without having the ability to drop packets. A spoofed packet will cause the real packet to be dropped by the endpoint if it arrives late.

**Comment by ekr** [comment]
> Because we [told it to](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#off-path-forward):
> 
> > In response to an apparent migration, endpoints MUST validate the previously active path using a PATH_CHALLENGE frame.

OK. I missed that section. However, I'm not sure how much it changes thing, though. The question is how much impact an on-path attacker can cheaply have (and of course, this goes back to what we claim our security guarantees are).

**Comment by martinthomson** [comment]
The properties we want from migration are distinct from those during the handshake.  I'd like to resolve that discussion first.  Then we can talk about what we might be willing to tolerate during migration.

**Comment by martinthomson** [comment]
Discussed in Tokyo: the working group believes right now that our existing defense is OK-ish.  Roberto noted that we're not current as good as multipath TCP and suggested that we might want stronger confirmation.

@ekr will take a first stab at documenting our principles with respect to what attacks we are prepared to spend effort to defend against.

**Comment by martinduke** [comment]
Section 9 has a SHOULD regarding PATH_CHALLENGE; change it to a MUST?

**Comment by martinthomson** [comment]
Proposal from Tokyo:

Endpoints MUST validate new paths before using them.  In reaction to a migration they MUST continue using the old path until they receive new non-probing packets on the new path AND validation of that path is complete.  

Research question: Allow an exception in the case that the rebinding appears to be a NAT rebinding.  Heuristics are necessary here: e.g., a port-only change after a period of inactivity.  If we adopt the strict requirement, but learn that we get a lot of hitches from apparent migration, then we can find a way to relax the requirement.  If we allow this exception, then we have to address the problem of what to do when there are multiple such changes in a short period.  ==> Open this as an issue to track the research question.

Otherwise, eliminate the distinction between IP and port changes and just say "address change".

Stricter rules would close off the ability for an attacker to force a connection into a non-functioning state, even if that state is temporary.  The cost is that migration can result in a delay of a round trip or more.

(We might also want to know more about the severity of the attack.)

**Comment by ianswett** [comment]
I'd add "appears to be intentional immediate migration" to the list of research questions.  If a PATH_CHALLENGE is bundled with other frames and the path changes, then it seems very likely the peer knows the path has changed and it's intentional.

**Comment by janaiyengar** [comment]
Chatting with @martinduke, we may also want to add notes about potential signals of attacks in the off-path forwarding case: (i) large number of duplicates are being received from multiple addresses, (ii) PATH_RESPONSE is received from multiple addresses. These may not be actionable for everyone, but it helps implementers understand interesting signatures.

**Comment by ianswett** [comment]
Q: Wouldn't the sender of a PATH_CHALLENGE ignore a PATH_RESPONSE that came back from a different peer address than what they send the PATH_CHALLENGE to?

But yes, duplicate packets are insanely rare and receiving a duplicate from different addresses would almost never happen without interference, so it should be a strong signal that the immediate migration should be ignored.  Of course that makes the state machine even more complex, but assuming that duplicate arrives while a PATH_CHALLENGE is outstanding, maybe it's not that much extra complexity?

**Comment by erickinnear** [comment]
We've gone through a bunch of different issues in this conversation, many of them have since been addressed. 

In general, we've come up with some requirements, like having both old and new paths be validated, removing many of the concerns around quiescence. There's also now a mention about heuristics, and it might be worth some editorial text describing possible heuristics to use (happy to put some of that together, will pretty much align with @martinduke's and @janaiyengar's comments above).

Much of @martinthomson's comment above with the Tokyo proposal still applies, some of the last bits of the text outlined there are changing in #2637. 

We can try to summarize the current state of things (also happy to do that), but I'm not sure that there's a lot left to do in this issue specifically.

**Comment by larseggert** [comment]
London discussion: @erickinnear to finalize

**Comment by mnot** [comment]
@erickinnear any progress here? Thanks.

**Comment by erickinnear** [comment]
@mnot Yes, just been blocked behind other things (like WWDC).

**Comment by erickinnear** [comment]
@ianswett To your question above, we don't validate the PATH_RESPONSE's address vs. the PATH_CHALLENGE destination anymore since #2637 (Stop requiring remote address to match during path validation).

**Comment by erickinnear** [comment]
Just to clarify what text I'm adding here, this is documenting our security stance towards attacks against migration, which things we think should be possible and which things we think should not be possible. (Which ekr can continue on into a section for QUIC in general.)

**Comment by erickinnear** [comment]
The PR (so far) establishes the following definitions and guarantees:

### On-Path Attacker
An on-path attacker can:
- Inspect packets
- Modify unencrypted packet headers
- Inject new packets
- Delay packets
- Drop packets

An on-path attacker cannot:
- Modify encrypted packet payloads

### Off-Path Attacker
An off-path attacker can:
- Inspect packets
- Inject new packets

An off-path attacker cannot:
- Modify any part of a packet
- Delay packets
- Drop packets

### Limited Off-Path Attacker
A limited on-path attacker can:
- Inspect packets
- Inject new packets
- Modify unencrypted packet headers

A limited on-path attacker cannot:
- Delay packets beyond the original packet duration
- Drop packets
- Modify encrypted packet payloads

## Guarantees
### On-Path Attacker
1. An on-path attacker can interrupt a QUIC connection, causing it to fail if it cannot migrate to a new path that does not contain the attacker. This can be achieved by dropping all packets, modifying them so that they fail to decrypt, or other methods.
2. An on-path attacker can prevent migration to a new path for which the attacker is also on-path by causing path validation to fail on the new path.
3. An on-path attacker cannot prevent a client from migrating to a path for which the attacker is not on-path.
4. An on-path attacker can reduce the throughput of a connection by delaying packets or dropping them.

### Off-Path Attacker
1. An off-path attacker can race packets and attempt to become a “limited” on-path attacker.
2. An off-path attacker can cause path validation to succeed for forwarded packets with the source address listed as the off-path attacker as long as it can provide improved connectivity between the client and the server.
3. An off-path attacker cannot cause a connection to close.
4. An off-path attacker cannot cause migration to a new path to fail if it cannot observe the new path.
5. An off-path attacker can become a limited on-path attacker during migration to a new path for which it is also an off-path attacker.
6. An off-path attacker can become a limited on-path attacker by affecting shared NAT state such that it sends packets to the server from the same IP address and port that the client originally used.

### Limited Off-Path Attacker
1. A limited on-path attacker cannot cause an active connection to close.
2. A limited on-path attacker cannot cause an idle connection to close if the client is first to resume activity.
3. A limited on-path attacker can cause an idle connection to be deemed lost if the server is the first to resume activity.

There's more descriptive text in the PR, as well, covering in a bit more detail our expectations about capabilities of the attackers, pathologically worst-case behaviors they may engage in, etc.

**Comment by mikkelfj** [comment]
I think there needs to be a distinction between observing off-path attackers and blind off-path attackers. A blind attacker can inject packets and for example try to force close a connection even if the traffic flow is not visible. Such an attacker is probably the most common.

**Comment by gorryfair** [comment]
Can we be careful with how we sue "off path" - I think this uses the term assuming some visibility of the packets. In many transport documents, we assume that an off-path attacker does not have a way to see the packets.

**Comment by erickinnear** [comment]
Yes, sorry @gorryfair, I've got a reference to RFC 3552 in there but haven't updated to fully use the correct terminology yet. Will do, and I'll ping you to take another look to make sure I get the right ones in the right places. :)

**Comment by erickinnear** [comment]
@gorryfair, I've now pushed some updates that try to make this clearer, another look and comments would be most welcome! (I expect this will take some iterating, but let's start with what's there and improve it.)

@mikkelfj I've added a paragraph about blind active attackers (which are probably off-path since it's hard to not see the packets if you're on-path). 

Overall, I've kept the distinction of on-path vs. off-path, and referenced the definition for such in RFC 3552 which may help make things a bit clearer.

**Comment by mikkelfj** [comment]
@erickinnear that makes sense. I just read the section on blind attacks. You mention it is not likely to successful which is probably true, but the reason is not limited to what you state. There is a lot of pain in path migration/discovery to avoid blind attacks being successful in hijacking or interrupting connections (I don't recall all the details).

**Comment by larseggert** [comment]
Discussed in Cupertino. Contributors to review #2925 ASAP.

**Comment by erickinnear** [comment]
@mikkelfj Happy to make changes, if you can recall some of the details here it would be great to be more explicit about them in the text. Otherwise, I think it does pretty much accurately reflect what we think is possible as an attack surface, so we should fold any concrete suggestions in ASAP.

**Comment by mikkelfj** [comment]
I think some of issues I was thinking about relate to a non-blind attack where packets are mirrored, but that might well be covered in your text. I'm just linking for reference

https://github.com/quicwg/base-drafts/issues/2582

Another issue that has been discussed is in early handshake where a powerful (government) adversary can race packets during handshake through a faster network. This is also not blind, and it is not something QUIC v1 attempts to solve AFAIR.

**Comment by janaiyengar** [comment]
@mnot @lars @LPardue : This should have been closed by #2925, except that this issue is marked as design and #2925 was marked as editorial. How do you want to proceed?
